### PR TITLE
fix(startup): ensure nucleus version config is always available

### PR DESCRIPTION
**Pull request by shaguptashaikh**
**Issue #, if available:**
If a config file is manually provided at startup with certain nucleus component config but no component version config(or other required keys for nucleus component), inferred version from build file or fallback version if even that's not present(which shouldn't happen in real usage) should be used instead of forcing user to include version in manually created config or finding out too late that it's needed
Also adds a null check in dependency resolution which depends on the version config and and fails without proper message if it's missing

**Description of changes:**

**Why is this change necessary:**
See description

**How was this change tested:**
Added unit tests

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [x] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/test/java/com/aws/greengrass/deployment/DeviceConfigurationTest.java [hunk 1]
```diff
@@ -73,9 +82,81 @@ void GIVEN_bad_cred_endpoint_config_WHEN_validate_THEN_fails() {
...
+            assertEquals("0.0.0", deviceConfiguration.getNucleusVersion());
+        } finally {
+            context.close();
+        }
+    }
+
+    @Test
+    void GIVEN_no_existing_config_WHEN_init_device_config_THEN_use_default_nucleus_config() throws Exception {
+        Context context = new Context();
+        try {
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/deployment/DeviceConfigurationTest.java:149]: use try with resources here and elsewhere

**Comment by MikeDombo** [comment]
To fix the commitlint, make the commit have a lowercase `Ensure`

**Review by MikeDombo — DISMISSED** [review]

**Review by hui-yang — APPROVED** [review]

**Review by MikeDombo — COMMENTED** [review]
Need to add logging of the deployment errors as well so that we can have proper logs for the next time this happens.

**Comment by shaguptashaikh** [comment]
> Need to add logging of the deployment errors as well so that we can have proper logs for the next time this happens.

I'm not sure which error logs you're referring to, I see that if errors like PackageDownloadException in this case are set as failure cause, the status detail already picks up the message. And DeploymentService also logs any unknown exceptions so I think we're covered. I'll send out a separate PR if something is still missing

**Comment by MikeDombo** [comment]
> > Need to add logging of the deployment errors as well so that we can have proper logs for the next time this happens.
> 
> I'm not sure which error logs you're referring to, I see that if errors like PackageDownloadException in this case are set as failure cause, the status detail already picks up the message. And DeploymentService also logs any unknown exceptions so I think we're covered. I'll send out a separate PR if something is still missing

We don't properly log here: https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/4b87b714b44ab837cedfb024b8207b6402716a16/src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java#L126

If we had logged, we could have identified the problem within seconds.

**Comment by MikeDombo** [comment]
If we merge Hui's change, do we still need this, or does her's take care of it?

**Comment by shaguptashaikh** [comment]
> If we merge Hui's change, do we still need this, or does her's take care of it?

I can rebase on that change but I want to get the tests from this PR in at least

**Review by hui-yang — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]

**Pull request merged**